### PR TITLE
Ensure TestPrincipalUser is thread safe

### PR DESCRIPTION
**Pull request by acodrington**
# Background

`SilentProcessRunnerFixture` appears to be a [flaky test fixture](https://build.octopushq.com/buildConfiguration/TeamFireAndMotion_OctopusTentacle_IntegrationTestNet60Windows2012/9511661?expandPull+Request+Details=true&expandBuildProblemsSection=true&hideProblemsFromDependencies=false&expandBuildTestsSection=true&hideTestsFromDependencies=false&expandBuildChangesSection=true), with various tests occasionally failing with the following error:

```
System.DirectoryServices.AccountManagement.PrincipalExistsException : The account already exists.

  ----> System.Runtime.InteropServices.COMException : The account already exists.
   at System.DirectoryServices.AccountManagement.SDSUtils.ApplyChangesToDirectory(Principal p, StoreCtx storeCtx, GroupMembershipUpdater updateGroupMembership, NetCred credentials, AuthenticationTypes authTypes)
   at System.DirectoryServices.AccountManagement.SDSUtils.InsertPrincipal(Principal p, StoreCtx storeCtx, GroupMembershipUpdater updateGroupMembership, NetCred credentials, AuthenticationTypes authTypes, Boolean needToSetPassword)
   at System.DirectoryServices.AccountManagement.SAMStoreCtx.Insert(Principal p)
   at System.DirectoryServices.AccountManagement.Principal.Save()
   at Octopus.Tentacle.Tests.Integration.Util.TestUserPrincipal..ctor(String username, String password) in C:\BuildAgent\work\639265b01610d682\source\Octopus.Tentacle.Tests.Integration\Util\TestUserPrincipal.cs:line 32
   at Octopus.Tentacle.Tests.Integration.Util.SilentProcessRunnerFixture.SetUpLocal() in C:\BuildAgent\work\639265b01610d682\source\Octopus.Tentacle.Tests.Integration\Util\SilentProcessRunnerFixture.cs:line 27
--COMException
   at System.DirectoryServices.DirectoryEntry.CommitChanges()
   at System.DirectoryServices.AccountManagement.SDSUtils.ApplyChangesToDirectory(Principal p, StoreCtx storeCtx, GroupMembershipUpdater updateGroupMembership, NetCred credentials, AuthenticationTypes authTypes)
```

Some investigation shows that [the setup method run before each test](https://github.com/OctopusDeploy/OctopusTentacle/blob/b81fa6a5725f2628e30cdf4759f8132b97cca7f7/source/Octopus.Tentacle.Tests.Integration/Util/SilentProcessRunnerFixture.cs#L24) attempts to create a Windows test user via the `TestUserPrincipal` class. The implementation seems to be **thread unsafe** - it first attempts to find an existing user, with a new user being created if an existing one cannot be found. It is entirely possible for these tests, when run in parallel, to not find an existing user then each try to create a new user, causing the observed error.

This PR wraps user access operations in a static lock, causing the class to be safer for multiple threads, and _hopefully_ addressing the flake.

# Results

🟢🤞

# Pre-requisites

- [x] I have read [How we use GitHub Issues](https://github.com/OctopusDeploy/Issues/blob/master/docs/CONTRIBUTING.internal.md) for help deciding when and where it's appropriate to make an issue.
- [x] I have considered informing or consulting the right people, according to the [ownership map](https://whimsical.com/ownership-map-NzbiD4HJyvhC9jNJNfS6TG).
- [x] I have considered appropriate testing for my change.

### CODE REVIEW

**Comment by acodrington** [comment]
[sc-59223]

File: source/Octopus.Tentacle.Tests.Integration/Util/TestUserPrincipal.cs [hunk 1]
```diff
@@ -50,12 +24,40 @@ public TestUserPrincipal(string username, string password = "Password01!")
             }
         }
 
+        static UserPrincipal GetUserPrincipal(string username, string password)
```
Comment (nathanwoctopusdeploy) [inline source/Octopus.Tentacle.Tests.Integration/Util/TestUserPrincipal.cs:27]: nit: GetExitingOrCreatePrincipal or similar  for clarity

**Comment by nathanwoctopusdeploy** [comment]
> class to be thread safe

I'm not sure you can say this as each call does change the password so it is still sot thread-safe, one thread could swap the password on another thread if they try and use different passwords.

**Review by nathanwoctopusdeploy — APPROVED** [review]

**Comment by acodrington** [comment]
@nathanwoctopusdeploy Would you accept "thread safer"? "thread-safe-ish"? 😄

**Pull request merged**